File: bindings/python/examples/ping.py
#!/usr/bin/env python3
import asyncio
import sys
import traceback
from lattica import Lattica

def parse_multiaddr(addr_str: str):
    if not addr_str.startswith('/'):
        raise ValueError(f"Invalid multiaddr format: {addr_str}")

    parts = addr_str.strip().split('/')
    if len(parts) < 6:
        raise ValueError(f"Incomplete multiaddr: {addr_str}")

    peer_id = None
    for i, part in enumerate(parts):
        if part == 'p2p' and i + 1 < len(parts):
            peer_id = parts[i + 1]
            break

    if not peer_id:
        raise ValueError(f"No peer ID found in multiaddr: {addr_str}")

    return addr_str, peer_id

async def print_peer_info(node, node_name):
    print(f"\n{node_name}:")
    peers = node.get_all_peers()
    print(f"  - peers: {len(peers)}")
    
    for peer_id in peers:
        peer_info = node.get_peer_info(peer_id)
        if peer_info:
            print(f"  - Peer {peer_id}")
            print(f"    - addresses: {peer_info.addresses}")
            print(f"    - RTT: {peer_info.rtt_ms}ms")
            print(f"    - last_seen: {peer_info.last_seen}")
            if peer_info.decay_3:
                print(f"    - decay_3: {peer_info.decay_3}ms")
            if peer_info.decay_10:
                print(f"    - decay_10: {peer_info.decay_10}ms")
            if peer_info.failures:
                print(f"    - failures: {peer_info.failures}")
            if peer_info.failure_rate:
                print(f"    - failure_rate: {peer_info.failure_rate}")

async def run_client():
    bootstrap_addr = sys.argv[1]
    bootstrap_addr, server_peer_id = parse_multiaddr(bootstrap_addr)
    bootstrap_nodes = [bootstrap_addr]

    client = Lattica.builder().with_bootstraps(bootstrap_nodes).build()

    await print_peer_info(client, "client")

    for i in range(1, 20):
        print(f"\nPing test {i}/5:")

        before_rtt = client.get_peer_rtt(server_peer_id)
        print(f"  - Before RTT: {before_rtt}s")

        # wait for next ping finish
        await asyncio.sleep(3)

        after_rtt = client.get_peer_rtt(server_peer_id)
        print(f"  - After RTT: {after_rtt}s")
        
        if before_rtt is not None and after_rtt is not None:
            diff = abs(after_rtt - before_rtt)
            print(f"  - RTT Change: {diff:.6f}s")

        peer_info = client.get_peer_info(server_peer_id)
        if peer_info:
            print(f"  - Current Server RTT: RTT={peer_info.rtt_ms}ms")
            addresses = client.get_peer_addresses(server_peer_id)
            print(f"  - Server address count: {len(addresses)}")

    print("\nfinal info:")
    await print_peer_info(client, "client")

async def run_server():
    server = Lattica.builder().build()

    while True:
        await print_peer_info(server, "server")
        await asyncio.sleep(10)

# node1: python ping.py
# node2: python ping.py /ip4/127.0.0.1/tcp/x/p2p/x
if __name__ == "__main__":
    try:
        if len(sys.argv) > 1:
            asyncio.run(run_client())
        else:
            asyncio.run(run_server())
    except KeyboardInterrupt:
        print("Shutting down...")
    except Exception as e:
        print(f"error: {e}")
        traceback.print_exc()

File: lattica/examples/ping.rs
use lattica::network::Lattica;
use lattica::common::extract_peer_id;
use libp2p::{Multiaddr, PeerId};
use std::time::Duration;
use tokio::time::sleep;
use std::env;

async fn print_node_info(node_name: &str, node: &Lattica) {
    println!("{}:", node_name);
    let peers = node.get_all_peers().await;
    println!("  - peers: {}", peers.len());
    for peer_id in &peers {
        if let Some(peer_info) = node.get_peer_info(peer_id).await {
            println!("  - Peer {}: RTT = {:?}", peer_id, peer_info.rtt());
            let addresses = node.get_peer_addresses(peer_id).await;
            println!("    - address: {:?}", addresses);
            println!("    - peer_info: {:?}", peer_info.clone());
        }
    }
}

// cargo run --example ping
// cargo run --example ping -- /ip4/127.0.0.1/tcp/x/p2p/x
#[tokio::main]
async fn main() -> anyhow::Result<()> {
    let args: Vec<String> = env::args().collect();
    let mut server_peer_id = PeerId::random();
    let mut bootstrap_nodes = vec![];
    if args.len() > 1 {
        let addr = args[1].parse::<Multiaddr>()?;
        bootstrap_nodes.push(addr.clone());

        if let Some(peer_id) = extract_peer_id(&addr) {
            server_peer_id = peer_id;
        }
    }

    if bootstrap_nodes.is_empty() {
        let mut node1 = Lattica::builder()
            // .with_listen_addrs(vec!["/ip4/127.0.0.1/tcp/0".parse().map_err(|e| anyhow::anyhow!("{}", e))?])
            .with_listen_addrs(vec!["/ip4/0.0.0.0/tcp/0".parse()?])
            // .with_listen_addrs(vec!["/ip4/0.0.0.0/udp/0/quic-v1".parse()?])
            // .with_listen_addr("/ip4/127.0.0.1/tcp/0".parse().map_err(|e| anyhow::anyhow!("{}", e))?)
            .with_mdns(true)
            .with_kad(true)
            .build()
            .await.map_err(|e| anyhow::anyhow!("{}", e))?;

        println!("node1 Peer ID: {}", node1.peer_id());

        loop {
            print_node_info("node1", &node1).await;
            sleep(Duration::from_secs(10)).await;
        }

    }else{
        // node2
        let mut node2 = Lattica::builder()
            // .with_listen_addrs(vec!["/ip4/127.0.0.1/tcp/0".parse().map_err(|e| anyhow::anyhow!("{}", e))?])
            .with_listen_addrs(vec!["/ip4/0.0.0.0/tcp/0".parse()?])
            // .with_listen_addrs(vec!["/ip4/0.0.0.0/udp/0/quic-v1".parse()?])
            // .with_listen_addr("/ip4/127.0.0.1/tcp/0".parse().map_err(|e| anyhow::anyhow!("{}", e))?)
            .with_mdns(true)
            .with_kad(true)
            .build()
            .await.map_err(|e| anyhow::anyhow!("{}", e))?;

        println!("node2 Peer ID: {}", node2.peer_id());

        let node1_addrs = bootstrap_nodes[0].clone();
        if node1_addrs.is_empty() {
            println!("node1 no address");
            return Ok(());
        }
        println!("node1 address: {:?}", node1_addrs);
        node2.dial(node1_addrs).await?;
        println!("\nbegin Ping test...");

        for i in 1..=5 {
            println!("Ping test {}: node2 ping node1", i);

            let before_rtt = node2.get_peer_rtt(&server_peer_id).await;
            println!("  - Ping before RTT: {:?}", before_rtt);

            sleep(Duration::from_secs(5)).await;

            let after_rtt = node2.get_peer_rtt(&server_peer_id).await;
            println!("  - Ping after RTT: {:?}", after_rtt);

            if let (Some(before), Some(after)) = (before_rtt, after_rtt) {
                let diff = if after > before {
                    after - before
                } else {
                    before - after
                };
                println!("  - RTT change: {:?}", diff);
            }

            sleep(Duration::from_secs(1)).await;
        }

        println!("\nPing done!");

        print_node_info("node2", &node2).await;
    }
    Ok(())
}